### ISSUE TITLE
[bug] Low Cosine Similarity in compare_hf_and_megatron due to attention_mask mismatch

### ISSUE THREAD

**Issue by hemaher0** [NVIDIA-NeMo/Megatron-Bridge#2466]
**Describe the bug**

There is a similarity incosistency in `examples/conversion/compare_hf_and_megatron/compare.py`, which is the example for conversion from hugging face model to megatron model.
And, I found that, this is because of how `attention_mask` is initialized and passed to the Megatron mode.

```py
def compare_models_one_step(args) -> None:
    """Compare 1-step generation between HF and Megatron models with debugging.

    Args:
        args: Parsed command line arguments
    """
# ==========================================================
# Skip the lines
# ==========================================================
    with torch.no_grad():
        position_ids = (
            torch.arange(input_ids.size(1), dtype=torch.long, device=input_ids.device)
            .unsqueeze(0)
            .expand_as(input_ids)
        )
        attention_mask = torch.ones_like(input_ids, dtype=torch.bool) # # <--- Here is the issue

        fwd_bwd_function = get_forward_backward_func()
        iterator = SingleBatchIterator(input_ids, position_ids, attention_mask, pixel_values, image_grid_thw)

        megatron_output = fwd_bwd_function(
            forward_step_func=vlm_forward_step,
            data_iterator=iterator,
            model=megatron_model,
            num_microbatches=1,
            forward_only=True,
            seq_length=input_ids.size(1),
            micro_batch_size=1,
            collect_non_loss_data=True,
        )
```
The fundamental issue lies in the semantic and structural differences of `attention_mask` between Hugging Face and Megatron-LM:
Hugging Face interprets True as tokens to attend to, whereas Megatron uses the mask to identify positions to mask out (by adding `-10000.0` to attention scores). Passing an all-True mask effectively blocks all tokens from attending to each other in Megatron.

**Steps/Code to reproduce bug**
1. Execute the comparison script examples/conversion/compare_hf_and_megatron/compare.py for a decoder-only/causal model.
```py
args.hf_model_path = "meta-llama/Meta-Llama-3-8B"
args.prompt = "Hi"
```
2. In the Megatron forward pass preparation, manually inject an HF-style mask: `attention_mask = torch.ones_like(input_ids, dtype=torch.bool)` instead of `None`.
3. Compare the outputs and observe that the cosine similarity drops significantly.

**Expected behavior**
The script should maintain a cosine similarity close to 1.0. To achieve this correct behavior for causal models, `attention_mask` should be explicitly set to `None` when calling the Megatron model. This allows Megatron's backend (e.g., FlashAttention or core modules) to accurately generate and apply the lower-triangular causal mask by itself.
Below are the comparisons of the outputs demonstrating the issue:

**Buggy Output (When injecting HF-style mask:** `attention_mask = torch.ones_like(...)`**)**
```md
Megatron output shape: torch.Size([1, 2, 128256])
Megatron logits stats - mean: -1.5639, std: 1.6396
Megatron next token: 931 ('000')
Megatron Top 5: [('000', 6.625), ('\u200b\u200b', 5.34375), (' realize', 5.1875), (' spend',
4.8125), ('!!!!!!!!', 4.78125)]
=== COMPARISON ===
Token match: False
Logits diff - max: 16.125000, mean: 2.677473
Cosine similarity: 0.601898
=== COMPARISON COMPLETE ===
```

**Corrected Output (When using** `attention_mask = None`**)**
```md
Megatron output shape: torch.Size([1, 2, 128256])
Megatron logits stats - mean: -2.9692, std: 2.7357
Megatron next token: 11 (',')
Megatron Top 5: [(',', 13.5625), ('!', 12.625), (' there', 12.3125), (' everyone', 12.125), (' all', 11.75)]
=== COMPARISON ===
Token match: True
Logits diff - max: 0.125000, mean: 0.019447
Cosine similarity: 1.000411
=== COMPARISON COMPLETE ===
```

**Additional context**
This issue often leads to false negatives when users try to verify the correctness of their Hugging Face to Megatron model conversions. Ensuring `attention_mask = None` for decoder-only architectures is essential for identical forward pass evaluation.

**Comment by yaoyu-33** [comment]
thanks for contribution, approved pr, triggered ci.

**Comment by pruprakash** [comment]
## QA RCCA Analysis

### 1. Fix Reference
- **PR**: #2476 - Fix attention_mask mismatch in compare.py
- **Issue**: #2466 - Low Cosine Similarity in compare_hf_and_megatron due to attention_mask mismatch

### 2. Root Cause
Low cosine similarity reported between HuggingFace and Megatron model outputs in `examples/conversion/compare_hf_and_megatron/compare.py`.

**Root cause**: `attention_mask` was initialized and passed differently to Megatron model vs HuggingFace model:
- HF model received proper causal attention mask
- Megatron model received mismatched or improperly formatted mask
- This caused numerical differences in attention computation

### 3. Trigger Configuration
- Run `compare.py` example to verify HF→Megatron conversion
- Observe low cosine similarity despite correct weight conversion
- Issue is in the comparison script, not the conversion itself

### 4. Nature of the Bug
**Classification**: CODE BUG - Example script bug
- Bug is in example/comparison script, not core library
- Conversion is correct, comparison methodology was wrong
- Caused false negatives in conversion validation

### 5. Existing Test Coverage
**Before Fix**: No automated test for compare.py script

**After Fix**: PR fixes the mask handling

### 6. Coverage Assessment
| Test Type | Exists | Covers Bug |
|-----------|--------|------------|
| Unit test for compare.py | NO | NO |
| Conversion roundtrip tests | YES | YES (compare weights, not outputs) |

### 7. Recommendation
The example script is for manual verification. Existing roundtrip tests that compare weights directly are more reliable. Low priority for additional tests.

### 8. Conclusion
**Verdict**: ADEQUATE COVERAGE (via weight comparison tests)

This was a bug in an example script used for manual verification. Core conversion tests compare weights directly which is more reliable than output comparison. The script fix is sufficient.
